### PR TITLE
Feature/series labels

### PR DESCRIPTION
**Pull request by robwebdev**
@msp I'm opening this for visibility, it's still WIP but feel free to comment.

- [x] Labels for lines on combo chart
- [x] Labels for line chart
- [x] Labels for bar chart 
- [x] Labels for bars on combo chart
- [x] Labels for pie chart segments

### CODE REVIEW

File: src/Data/Widget/Adapters/ChartAdapter.elm [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+                            |> CellRange.extractRows data
+                            |> List.map (\a -> Maybe.withDefault "" (List.head a))
+                in
+                    Just labels
+
+            Nothing ->
+                Nothing
+
+
+adapt : AdapterConfig.Config -> Data -> ( Row, List Row, Float, Float, Row, Maybe (List String) )
```
Comment (msp) [inline src/Data/Widget/Adapters/ChartAdapter.elm]: Can we introduce Chart.Data here?

File: src/Data/Widget/Adapters/TableAdapter.elm [hunk 2]
```diff
@@ -60,6 +60,9 @@ adapt optionalConfig data =
         rowToNumber row =
             List.map NumberParser.fromString row
 
+        rowToInt row =
```
Comment (msp) [inline src/Data/Widget/Adapters/TableAdapter.elm]: Unused?

**Review by msp — COMMENTED** [review]
Looking good! Added a couple of minor comments.

**Comment by robwebdev** [comment]
This is ready for another review when you're ready @msp

File: src/Views/Widget/Renderers/BarChart.elm [hunk 3]
```diff
@@ -17,18 +26,16 @@ import Views.Widget.Renderers.Config as ViewConfig
...
-        TABLE optionalConfig ->
+        CHART optionalConfig ->
             let
-                ( headerRow, bodyRows, minValue, maxValue, xLabels ) =
-                    TableAdapter.adapt optionalConfig data
-
-                dataAsHeaderValueTuples =
-                    List.map (List.map2 (,) headerRow) bodyRows
+                chartData =
+                    ChartAdapter.adapt optionalConfig data
```
Comment (msp) [inline src/Views/Widget/Renderers/BarChart.elm:38]: Nice

File: src/Views/Widget/Renderers/ChartLegend.elm [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+renderHorizontalLabel index labelText labelPrefix getLabelColor =
+    renderLabel index labelText labelPrefix getLabelColor [ Svg.Attributes.dx "10" ]
+
+
+renderVerticalLabel : Int -> String -> String -> (Int -> String) -> Svg msg
+renderVerticalLabel index labelText labelPrefix getLabelColor =
+    renderLabel index labelText labelPrefix getLabelColor [ Svg.Attributes.dy "20", Svg.Attributes.x "10" ]
+
+
+renderLabels : Maybe (List String) -> (Int -> String -> Svg msg) -> List (Svg msg)
```
Comment (msp) [inline src/Views/Widget/Renderers/ChartLegend.elm]: Minor comment: are we ever likely to call this direct, with a subsequent call to `ChartLegend.render`? Just wonder whether it should be called `createLabels` or something? Just thinking of how the API reads on use..

I see that we _could_ use it directly though! :)

Comment (msp) [reply src/Views/Widget/Renderers/ChartLegend.elm]: Same with its variants `renderHorizontalLabel` etc I suppose..

File: src/Views/Widget/Renderers/LineChart.elm [hunk 5]
```diff
@@ -225,6 +227,23 @@ renderYAxis width height continuousScale opts =
...
+
+
+renderLegend :
+    Int
+    -> Maybe (List String)
+    -> List (Svg msg)
+renderLegend top seriesLabels =
+    let
+        labels =
+            ChartLegend.renderLabels seriesLabels renderLegendLabel
```
Comment (msp) [inline src/Views/Widget/Renderers/LineChart.elm]: `ChartLegend.createLabels seriesLabels legendLabel` ?

File: src/Views/Widget/Renderers/LineChart.elm [hunk 6]
```diff
@@ -225,6 +227,23 @@ renderYAxis width height continuousScale opts =
             [ yAxis ]
 
 
+renderLegendLabel : Int -> String -> Svg msg
+renderLegendLabel index labelText =
+    ChartLegend.renderHorizontalLabel index labelText "―" getLineColour
```
Comment (msp) [inline src/Views/Widget/Renderers/LineChart.elm]: `ChartLegend.createHorizontalLabel index labelText "―" getLineColour` ?

**Review by msp — COMMENTED** [review]
Nice one @robwebdev ! Overall, really impressed with how quickly you picked this stuff up. Good work :)

Some minor discussion points on naming but it's just that.. a discussion so LMK your thoughts.

Only real niggle is the way the legends render on line/bar. Need to look again with fresh eyes but not sure if they need more top margin or centering. Or something. What do you think? I know the latter might be tricky..

**Comment by msp** [comment]
Just messing about: 

screenshot 2018-02-11 22 25 50

vs

screenshot 2018-02-11 22 26 38

File: src/Views/Widget/Renderers/ChartLegend.elm (hunk 4 above)
Comment (robwebdev) [reply src/Views/Widget/Renderers/ChartLegend.elm]: Yeah I agree, `create` does read nicer. My instinct was that if it's returning `Svg` (or a list of) then it's rendering, but happy to make those changes.

**Comment by robwebdev** [comment]
Hey @msp. Yeah I agree the positioning was a bit whack. In the latest commit I've split the `ChartLegend.render` function into `ChartLegend.renderTopLeftAligned` which is used by `PieChart` and `ChartLegend.renderBottomCenterAligned` which is used by the rest of the charts.

File: src/Views/Widget/Renderers/PieChart.elm [hunk 7]
```diff
@@ -111,7 +111,7 @@ view width height data =
                 data
 
         labels =
-            List.map Tuple.first data
+            List.map (\d -> Tuple.first d ++ ": " ++ Tuple.second d) data
```
Comment (robwebdev) [inline src/Views/Widget/Renderers/PieChart.elm:114]: Good shout!

**Pull request merged**